File: src/test/java/com/example/tarefas/controller/CategoriaControllerTest.java
package com.example.tarefas.controller;

import com.example.tarefas.model.Categoria;
import com.example.tarefas.repository.CategoriaRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class CategoriaControllerTest {

    @Mock
    private CategoriaRepository categoriaRepository;

    @InjectMocks
    private CategoriaController categoriaController;

    @Test
    void deveListarTodasAsCategorias() {
        Categoria c1 = new Categoria();
        c1.setDescricao("Trabalho");
        Categoria c2 = new Categoria();
        c2.setDescricao("Estudos");

        when(categoriaRepository.findAll()).thenReturn(List.of(c1, c2));

        List<Categoria> resultado = categoriaController.listarTodas();

        assertEquals(2, resultado.size());
        assertEquals("Trabalho", resultado.get(0).getDescricao());
        verify(categoriaRepository, times(1)).findAll();
    }

    @Test
    void deveCriarCategoriaComSucesso() {
        Categoria categoria = new Categoria();
        categoria.setDescricao("Nova Categoria");

        when(categoriaRepository.save(any(Categoria.class))).thenReturn(categoria);
        Categoria resultado = categoriaController.criar(categoria);

        assertNotNull(resultado);
        assertEquals("Nova Categoria", resultado.getDescricao());
        verify(categoriaRepository, times(1)).save(categoria);
    }
}

File: src/test/kotlin/com/example/tarefas/controller/CategoriaControllerTest.kt
package com.example.tarefas.controller

import com.example.tarefas.model.Categoria
import com.example.tarefas.repository.CategoriaRepository
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers.any
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull


@ExtendWith(MockitoExtension::class)
class CategoriaControllerTest {

    @Mock
    private lateinit var categoriaRepository: CategoriaRepository

    @InjectMocks
    private lateinit var categoriaController: CategoriaController

    @Test
    fun deveListarTodasAsCategorias() {
        val c1 = Categoria(descricao = "Trabalho")
        val c2 = Categoria(descricao = "Estudos")

        `when`(categoriaRepository.findAll()).thenReturn(listOf(c1, c2))

        val resultado = categoriaController.listarTodas()

        assertEquals(2, resultado.size)
        assertEquals("Trabalho", resultado[0].descricao)
        verify(categoriaRepository, times(1)).findAll()
    }

    @Test
    fun deveCriarCategoriaComSucesso() {
        val categoria = Categoria(descricao = "Nova Categoria")

        `when`(categoriaRepository.save(any(Categoria::class.java))).thenReturn(categoria)

        val resultado = categoriaController.criar(categoria)

        assertNotNull(resultado)
        assertEquals("Nova Categoria", resultado.descricao)
        verify(categoriaRepository, times(1)).save(categoria)
    }
}
